### PR TITLE
Dockerfile improvements

### PR DESCRIPTION
**Pull request by cfergeau**
# Proposed Changes

This PR does various improvements to the Dockerfile:
- removes unneeded package, which reduces the image size by about 10%
- run apt-get upgrade to get the latest security updates
- ~~improve the way package version are specified in apt-get install so that this line does not have to be updated every time one of these packages is updated~~

## Related Issues

> ([Github link][autolink-references] to related issues or pull requests)

[autolink-references]: https://help.github.com/articles/autolinked-references-and-urls/

### CODE REVIEW

File: unifi/Dockerfile [hunk 1]
```diff
@@ -8,13 +8,13 @@ SHELL ["/bin/bash", "-o", "pipefail", "-c"]
 # Setup base system
 RUN \
     apt-get update \
+    && apt-get upgrade -y \
```
Comment (frenck) [inline unifi/Dockerfile:11]: This should not be run here. This is the responsibility of the base image.

```suggestion
```

File: unifi/Dockerfile [hunk 2]
```diff
@@ -8,13 +8,13 @@ SHELL ["/bin/bash", "-o", "pipefail", "-c"]
...
     && apt-get install -y --no-install-recommends \
-        binutils=2.34-6ubuntu1.7 \
-        jsvc=1.0.15-8 \
-        libcap2=1:2.32-1 \
-        logrotate=3.14.0-4ubuntu3 \
+        binutils=2.34-* \
+        libcap2=1:2.32-* \
+        logrotate=3.14.* \
```
Comment (frenck) [inline unifi/Dockerfile:15]: Please restore exact version pinning. We have an bot upgrading them.
Also, this messes with Docker caching.

**Review by frenck — CHANGES_REQUESTED** [review]

File: unifi/Dockerfile (hunk 1 above)
Comment (cfergeau) [reply unifi/Dockerfile:11]: In general I agree, in this specific case, the 8.2.0 base image we are using is 100MB and 1 year old, there's a 8.2.1 base image, but it's also 1 year old and 500MB (5 times as big).
To get newer images, we'd have to switch to a 9.0.x base image, but it did not look like a simple change.

`apt-get upgrade -y` felt like an acceptable trade-off wrt security until the add-on is moved to 9.0.x.
I'll drop the commit if you disagree.

File: unifi/Dockerfile (hunk 2 above)
Comment (cfergeau) [reply unifi/Dockerfile:15]: Ah ok, was not aware of the bot, and saw some build issues related to that in another PR. I'll drop it.

**Comment by frenck** [comment]
> run apt-get upgrade to get the latest security updates

I'm not aware there is one? If so, please let me know what is going on, and let's fix it in the base image.

Upgrading packages on this level is also pretty inefficient.

> removes unneeded package, which reduces the image size by about 10%

You've removed `jsvc`, which is assumingly used internally by UniFi, see also their installer scripts by the community linked to from UniFi: <https://get.glennr.nl/unifi/install/unifi-8.0.26.sh>

I don't think that is correct? Can you provide the reasoning for this removal and where I can information on this otherwise?

> improve the way package version are specified in apt-get install so that this line does not have to be updated every time one of these packages is updated

I was not aware this was a problem in this repo? Could you clarify where and why this is an issue?

File: unifi/Dockerfile (hunk 2 above)
Comment (cfergeau) [reply unifi/Dockerfile:15]: NB: current `main` is missing this update:
```
diff --git a/unifi/Dockerfile b/unifi/Dockerfile
index 9398a12..9aef07d 100755
--- a/unifi/Dockerfile
+++ b/unifi/Dockerfile
@@ -10,7 +10,7 @@ RUN \
     apt-get update \
     && apt-get upgrade -y \
     && apt-get install -y --no-install-recommends \
-        binutils=2.34-6ubuntu1.7 \
+        binutils=2.34-6ubuntu1.8 \
         libcap2=1:2.32-1 \
         logrotate=3.14.0-4ubuntu3 \
         mongodb-server=1:3.6.9+really3.6.8+90~g8e540c0b6d-0ubuntu5.3 \
```

Comment (cfergeau) [reply unifi/Dockerfile]: Dropped this commit from the PR.

File: unifi/Dockerfile (hunk 1 above)
Comment (renewoensdregt) [reply unifi/Dockerfile:11]: > To get newer images, we'd have to switch to a 9.0.x base image, but it did not look like a simple change.

I had looked at this too to go to later versions of the base image. I think the main issue is the fact that there is no package for Mongodb for Jammy. For all other main packages there are equivalents in Jammy.

**Comment by cfergeau** [comment]
> I'm not aware there is one? If so, please let me know what is going on, and let's fix it in the base image.

Running `apt-get upgrade` in `ghcr.io/hassio-addons/ubuntu-base:8.2.0` gives:
```
apt base-files bsdutils ca-certificates curl fdisk gcc-10-base libapt-pkg6.0 libasn1-8-heimdal libblkid1 libc-bin libc6 libcurl4 libfdisk1 libgcc-s1 libgnutls30 libgssapi-krb5-2 libgssapi3-heimdal libhcrypto4-heimdal libheimbase1-heimdal libheimntlm0-heimdal libhx509-5-heimdal libk5crypto3 libkrb5-26-heimdal libkrb5-3 libkrb5support0 libmount1 libncurses6 libncursesw6 libnghttp2-14 libpam-modules libpam-modules-bin libpam-runtime libpam0g libprocps8 libroken18-heimdal libsmartcols1 libsqlite3-0 libssh-4 libssl1.1 libstdc++6 libsystemd0 libtinfo6 libudev1 libuuid1 libwind0-heimdal mount ncurses-base ncurses-bin openssl perl-base procps tar tzdata util-linux
```
which is a fairly large set of packages. https://github.com/hassio-addons/addon-unifi/pull/481#discussion_r1453679837 has  more details why I could not find a better base image to use instead. Or maybe official addon builds are using automatically built base image or something like this, and not `ghcr.io/hassio-addons/ubuntu-base:8.2.0` as specified in the Dockerfile?

**Comment by cfergeau** [comment]
> You've removed jsvc, which is assumingly used internally by UniFi, see also their installer scripts by the community linked to from UniFi: https://get.glennr.nl/unifi/install/unifi-8.0.26.sh

> I don't think that is correct? Can you provide the reasoning for this removal and where I can information on this otherwise?

I've added some reasoning in the commit log:
```
The unifi package does not depend on it, `git grep jsvc` does not return a match, this is the only binary the jsvc package provides, and jsvc pulls openjdk-jre-11 while the rest of the code switched to 17
```
"does not depend on it" in the sense of "the unifi package does not have a dpkg/apt dependency for it"
I've also checked I can start the image without it (I did not try to go further than the first unifi screen on port 8443)

**Comment by cfergeau** [comment]
> I was not aware this was a problem in this repo? Could you clarify where and why this is an issue?

Not a huge issue if there's a bot automatically updating the repo, but for example this PR is not buildable because `main` is missing
```
-        binutils=2.34-6ubuntu1.7 \
+        binutils=2.34-6ubuntu1.8 \
```

**Comment by cfergeau** [comment]
> You've removed jsvc, which is assumingly used internally by UniFi, see also their installer scripts by the community linked to from UniFi: https://get.glennr.nl/unifi/install/unifi-8.0.26.sh

I've looked closer at this script, for unifi 8.x it has:
```
if [[ "${first_digit_unifi}" -gt '8' ]] || [[ "${first_digit_unifi}" == '8' && "${second_digit_unifi}" -ge "1" ]]; then
  ...
  required_java_version="openjdk-17"
  required_java_version_short="17"
fi
```
and jsvc installation is only done for openjdk-8, not for newer openjdk versions:
```
  if [[ "${required_java_version}" == "openjdk-8" ]]; then
    echo -e "${WHITE_R}#${RESET} Installing jsvc and libcommons-daemon-java..."
    if DEBIAN_FRONTEND='noninteractive' apt-get -y "${apt_options[@]}" -o Dpkg::Options::='--force-confdef' -o Dpkg::Options::='--force-confold' install jsvc libcommons-daemon-java &>> "${eus_dir}/logs/apt.log"; then echo -e "${GREEN}#${RESET} Successfully installed jsvc and libcommons-daemon-java! \\n"; else echo -e "${RED}#${RESET} Failed to install jsvc and libcommons-daemon-java in the first run...\\n"; unifi_dependencies=fail; fi
  fi
```

I also found https://community.ui.com/questions/New-UniFi-Controller-Installation-Issues-Ubuntu-18-04/a4a3b308-0ed4-4a4a-b54d-e0b59f9e4661 which indicates that at some point in the past, the unifi debian package had a requirement the jsvc package: `The following packages have unmet dependencies:  unifi : Depends: jsvc (>= 1.0.8) but it is not installable`. This dependency is not present in the latest version of the package.

File: unifi/Dockerfile (hunk 1 above)
Comment (cfergeau) [reply unifi/Dockerfile:11]: > I had looked at this too to go to later versions of the base image. I think the main issue is the fact that there is no package for Mongodb for Jammy. For all other main packages there are equivalents in Jammy.

[offtopic for this PR]

I actually gave it a try, wip branch using https://get.glennr.nl/unifi/install/unifi-8.0.26.sh for inspiration can be found at https://github.com/cfergeau/addon-unifi/tree/base9
The addon fails to start, but it's hopefully something minor, as it is working if I start the container with `--entrypoint=/bin/bash` and then run `/etc/init.d/unifi start`

**Review by frenck — APPROVED** [review]
Thanks, @cfergeau 👍

**Pull request merged**